### PR TITLE
os/board/rtl8730e: Support CMSIS-DSP

### PR DESCRIPTION
**Pull request by ZhenBei-Sin**
[build/configs/rtl8721csm: Clean up CMSIS-DSP]
[os/board/rtl8730e: Clean Up Makefile]
[configs/rtl8730e/flat_dev: Enable CMSIS-DSP by default]
[external/include/cmsis_nn: Move CMSIS-NN header files to external]
[external/include/cmsis_dsp: Move CMSIS-DSP header files to external]
[external/cmsis_dsp: Support CMSIS-DSP for rtl8730e]

### CODE REVIEW

**Comment by edwakuwaku** [comment]
Hi @ZhenBei-Sin , please add code size usage when applying cmsis-dsp library

**Comment by ZhenBei-Sin** [comment]
> Hi @ZhenBei-Sin , please add memory usage (ie. both flash and ram) when applying cmsis-dsp library

The library's memory usage change when its functions are invoked or called by an application.
So, it should be take care by Samsung side.

Please refer for the below result:

flat_apps without CMSIS-DSP
========== Size Verification of built Binaries ==========
Type        Binary Size     Partition Size      used(%)
 KERNEL      1,500,402 bytes    7,204,864 bytes     20.82%    PASS
=> Size verification SUCCESS!! The size of all binaries are OK.


flat_apps with Enable External to Link CMSIS-DSP
KERNEL size NO changes.
External Libraries >  [*] CMSIS DSP library 
========== Size Verification of built Binaries ==========
Type        Binary Size     Partition Size      used(%)
 KERNEL      1,500,402 bytes    7,204,864 bytes     20.82%    PASS
=> Size verification SUCCESS!! The size of all binaries are OK.


flat_apps with Enable External to Link CMSIS-DSP & Enable CMSIS-DSP example
External Libraries >  [*] CMSIS DSP library 
Application Configuration > Examples > Board Specific Demos > [*] CMSIS DSP example
========== Size Verification of built Binaries ==========
Type        Binary Size     Partition Size      used(%)
 KERNEL      1,595,770 bytes    7,204,864 bytes     22.15%    PASS
=> Size verification SUCCESS!! The size of all binaries are OK

**Review by edwakuwaku — APPROVED** [review]
LGTM.

File: apps/examples/board_specific_demo/ameba_cmsis_dsp_demo/README.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@ examples/cmsis_dsp
 ^^^^^^^^^^^^^^
 
   cmsis_dsp APIs is used to run the unit test provided by ARM.
-  This is the CMSIS_DSP example for rtl8721cms.
+  This is the CMSIS_DSP example for rtl8721cms and rtl8730e.
```
Comment (sunghan-chang) [inline apps/examples/board_specific_demo/ameba_cmsis_dsp_demo/README.txt:5]: cms -> csm

**Comment by sunghan-chang** [comment]
.a is in the external, but headers are in arch. Is that a part of arch or external? The arch is kernel but the external is user.

**Comment by sunghan-chang** [comment]
[build/configs/rtl8730e: Add config for support CMSIS-DSP](https://github.com/Samsung/TizenRT/pull/6261/commits/7259166ea386ac9f3afa12455ff8bbddebefcbf8)
I can't get what is changed from the title. How about `configs/rtl8730e/flat_dev: enable CMSIS-DSP by default`?
Even the commit has changes in other defconfigs, it just adds disabled (no operation change). So above would be ok I think.

**Comment by ZhenBei-Sin** [comment]
> .a is in the external, but headers are in arch. Is that a part of arch or external? The arch is kernel but the external is user.

Hi @sunghan-chang ,
Due to the starting phase for support cmsis, I put the cmsis headers file under the arch folder.
Recently, I was referred to Samsung's team members commit to moving it to the External path for support cmsis.

For better maintenance, yes. We should move it to be under the External path.
But...... May Samsung advise on howthe link or method to include the header file located at the External path for flat_apps and loadable_apps for the example.
#include <> ?

Thank You.

**Comment by sunghan-chang** [comment]
> > .a is in the external, but headers are in arch. Is that a part of arch or external? The arch is kernel but the external is user.
> 
> Hi @sunghan-chang , Due to the starting phase for support cmsis, I put the cmsis headers file under the arch folder. Recently, I was referred to Samsung's team members commit to moving it to the External path for support cmsis.
> 
> For better maintenance, yes. We should move it to be under the External path. But...... May Samsung advise on howthe link or method to include the header file located at the External path for flat_apps and loadable_apps for the example. #include <> ?
> 
> Thank You.

https://github.com/Samsung/TizenRT/blob/master/build/configs/rtl8730e/Make.defs#L82C3-L85C106
You can find the include paths here. 
For example, to include https://github.com/Samsung/TizenRT/blob/master/external/include/mbedtls/aes.h, you can do as below.
```
#include <mbedtls/aes.h>
```
`external/include/` path is in the Make.defs so that you can write remain path in the including.

If you add cmsis-dsp headers like external/include/cmsis-dsp/amebasmart/aa.h, we can include it as below
```
#include <cmsis-dsp/amebasmart/aa.h>
```

But I don't know the chipset name is necessary in the path.

The including is the same for both, the flat and the loadable.
But the external is user context so that kernel (os folder) should not call the function.

**Comment by ZhenBei-Sin** [comment]
> > > .a is in the external, but headers are in arch. Is that a part of arch or external? The arch is kernel but the external is user.
> > 
> > 
> > Hi @sunghan-chang , Due to the starting phase for support cmsis, I put the cmsis headers file under the arch folder. Recently, I was referred to Samsung's team members commit to moving it to the External path for support cmsis.
> > For better maintenance, yes. We should move it to be under the External path. But...... May Samsung advise on howthe link or method to include the header file located at the External path for flat_apps and loadable_apps for the example. #include <> ?
> > Thank You.
> 
> https://github.com/Samsung/TizenRT/blob/master/build/configs/rtl8730e/Make.defs#L82C3-L85C106 You can find the include paths here. For example, to include https://github.com/Samsung/TizenRT/blob/master/external/include/mbedtls/aes.h, you can do as below.
> 
> ```
> #include <mbedtls/aes.h>
> ```
> 
> `external/include/` path is in the Make.defs so that you can write remain path in the including.
> 
> If you add cmsis-dsp headers like external/include/cmsis-dsp/amebasmart/aa.h, we can include it as below
> 
> ```
> #include <cmsis-dsp/amebasmart/aa.h>
> ```
> 
> But I don't know the chipset name is necessary in the path.
> 
> The including is the same for both, the flat and the loadable. But the external is user context so that kernel (os folder) should not call the function.

Hi @sunghan-chang ,

Thank you for providing the useful information and guidance. 
I made the changes based on your suggestions and verified that it works well.
CMSIS header files for amebad and amebasmart are same, so will be remain one only.

Thank You :100:

**Comment by sunghan-chang** [comment]
@vibhor-m @Taejun-Kwon Could you check this PR for CMSIS-DSP?

**Comment by vibhor-m** [comment]
> @vibhor-m @Taejun-Kwon Could you check this PR for CMSIS-DSP?

external/include/cmsis_dsp/PrivateInclude can be moved to external/cmsis_dsp/include/PrivateInclude.
Header file mentions it private to library and IMO should not be exposed.

**Comment by ZhenBei-Sin** [comment]
> > @vibhor-m @Taejun-Kwon Could you check this PR for CMSIS-DSP?
> 
> external/include/cmsis_dsp/PrivateInclude can be moved to external/cmsis_dsp/include/PrivateInclude. Header file mentions it private to library and IMO should not be exposed.

Based on the suggestion, moved to the external/cmsis_dsp/include/PrivateInclude.
Done. Please review.

Thank You

**Review by Taejun-Kwon — APPROVED** [review]

**Comment by Taejun-Kwon** [comment]
Test finished, no issue

**Pull request merged**